apps/web/src/app/(profile)/[username]/_components/activity-chart.tsx

```diff
@@ -0,0 +1,167 @@
+
+export function ActivityChart(props: {
+  data: {
+    date: Date;
+    day: 0 | 1 | 2 | 3 | 4 | 5 | 6;
+    week: number;
+    month: number;
+    comments: number;
+    badges: number;
+    submissions: number;
+    activity: number;
+  }[];
+}) {
+  return (
+    <ChartContainer config={chartConfig} className="pointer-events-auto aspect-[8/7] h-[270px]">
+      <ScatterChart data={props.data} accessibilityLayer>
+        <ChartTooltip
+          cursor={false}
+          content={({ payload, content, ...props }) => {
+            const innerPayload = (payload?.[0]?.payload as Record<string, number>) ?? {};
+            const customPayload = Object.entries(innerPayload ?? {})
+              .filter(([key]) => key === 'comments' || key === 'badges' || key === 'submissions')
+              .map(([key, val]) => ({ name: key, value: val, payload }));
+            return <ChartTooltipContent {...props} payload={customPayload} />;
+          }}
+          labelFormatter={(_, [val]) => {
+            return format(val?.payload[0].payload.date, 'dd MMM');
+          }}
+        />
+        <XAxis
+          orientation="bottom"
+          dataKey="week"
+          type="category"
+          allowDuplicatedCategory={false}
+          tickLine={false}
+          axisLine={false}
+          tickFormatter={(val, idx) => {
+            return getMonthFromWeek(val, idx);
+          }}
+        />
+
+        <YAxis dataKey="day" type="category" allowDuplicatedCategory={false} hide />
+        <ZAxis dataKey="activity" type="number" />
+        <Scatter
+          // ReCharts has this type defined as Record<string, any> so attempting
+          // to overide it throws an error
+          shape={(item: unknown) => getShape(item)}
+        />
+      </ScatterChart>
+    </ChartContainer>
+  );
+}
+const getShape = (item: unknown) => {
+  const { cx, cy, activity } = item as {
+    cx: number;
+    cy: number;
+    activity: number;
+  };
+  const squareLength = 24;
+  const borderV = 10;
+  const borderH = 10;
+  return (
+    <SquareWrapper>
+      {(isSelected) => (
+        <>
+          <rect
+            x={cx - squareLength / 2}
+            y={cy - squareLength / 2}
+            width={squareLength}
+            height={squareLength}
+            className={cn(getColor(activity), isSelected ? 'stroke-primary stroke-2' : '')}
+            rx={2}
+          />
+          <rect
+            x={cx - borderH / 2 - squareLength / 2}
+            y={cy - borderV / 2 - squareLength / 2}
+            width={squareLength + borderH}
+            height={squareLength + borderV}
+            className="fill-transparent"
+          />
+        </>
+      )}
+    </SquareWrapper>
+  );
+};
+
+const activeNodeAtom = atom<PrimitiveAtom<boolean> | null>(null);
+const SquareWrapper = (props: { children: (isSelected: boolean) => React.ReactElement }) => {
+  const baseSelectedAtom = useMemo(() => atom(false), []);
+  const selectedAtom = useMemo(
+    () =>
+      atom(
```

Comment:
hmm should this atom not be defined in a component body?

Reply:
I don't think so but I'm down to hear other implementation methods (can I memo the component and get something similar ?).

 I did it this way, since I need an atom for each of the boxes. This atom is then put inside the global activeNodeAtom when it is clicked. Allows us to keep track of the atom of the currently clicked box. 

Then during the next click, we set the atom inside of the activeNodeAtom to false, and update that reference to the newly clicked atom. 

This way i can avoid re-rendering every single box on click.